src/main/docs/guide/appendix/architecture/aopArch.adoc

```diff
@@ -0,0 +1,123 @@
+    private final ExecutableMethod[] $proxyMethods = new ExecutableMethod[1];
+
+    public $NotNullExample$Definition$Intercepted(
+        BeanResolutionContext resolution, 
+        BeanContext context, 
+        Qualifier qualifier, 
+        List<Interceptor> interceptors) {
+        Exec executableMethods = new Exec(true); // <2>
+        this.$proxyMethods[0] = executableMethods.getExecutableMethodByIndex(0); // <3>
+        this.$interceptors[0] = InterceptorChain
+            .resolveAroundInterceptors(
+                context, 
+                this.$proxyMethods[0], 
+                interceptors
+        );  // <4>
+    }
+}
+----
+
+<1> The `@Generated` subclass extends from the decorated type and implements the api:aop.Interepted[] interface
+<2> An instance of api:context.ExecutableMethodsDefinition[] is constructed to resolve reflection-free dispatchers to the original method.
+<3> An internal array called `$proxyMethods` holds a reference for to each api:inject.ExecutableMethod[] instance used to proxy the invocation.
+<4> An internal array called `$interceptors` holds references to which api:aop.Interceptor[] instances apply to each method since an ann:aop.InterceptorBinding[] can be type or method level these may differ for each method.
+
+Each non-final and non-private method of the proxied type that has an ann:aop.InterceptorBinding[] associated with it (either type level or method level) is overridden with logic that proxies the original method, for example:
+
+[source,java]
+----
+@Overrides
+public void doWork(String taskName) {
+    ExecutableMethod method = this.$proxyMethods[0];
+    Interceptor[] interceptors = this.$interceptors[0]; // <1>
+    MethodInterceptorChain chain = new MethodInterceptorChain( // <2>
+        interceptors, 
+        this, 
+        method, 
+        new Object[]{taskName}
+    );
+    chain.proceed(); // <3>
+}
+----
+
+<1> The api:inject.ExecutableMethod[] and array of api:aop.Interceptor[] instances for the method is located.
+<2> A new api:aop.chain.MethodInterceptorChain[] is constructed with the interceptors, a reference to the intercepted instance, the method and the arguments.
+<3> The `proceed()` method is invoked on the api:aop.chain.MethodInterceptorChain[].
+
+Note that the default behaviour of the the ann:aop.Around[] annotation is to invoke methods on the intercepted super type that is subclassed so there is effectively only a single instance of the `NotNullExample` in the example above.
+
+However, this behaviour can be customized, for example using `@Around(proxyTarget=true)` will result two api:inject.BeanDefinition[] instances, one of the proxy and one for the proxy target. In this case the proxy becomes the primary bean and lookups up the proxy target and there are effectively two instances of the type.
+
+The default behaviour (`proxyTarget=false`) is more efficient memory wise as only a single api:inject.BeanDefinition[] is required.
```

Comment:
This whole section, the proxy and AOP section, is quite difficult to grasp. I'm not sure I understood all properly. I think what would help is a diagram with the different types and generated types for the example, which shows what is going on when a method is _invoked_. Also possibly variation with `proxyTarget=true`.

Reply:
I rewrote it and added a diagram, can you confirm it makes more sense?